2022/day7/day7.hs

```diff
@@ -0,0 +1,132 @@
+
+main = do
+  -- https://adventofcode.com/2022/day/7/input
+  commandLog <- readFile "input.log"
+
+  -- Process the log to build the list of directories with their total sizes
+  -- Flatten the ancestors to produce a full (flattened) tree of directories and sizes
+  let dirEntries = flattenAncestors (parentEntries (processLog commandLog))
+
+  -- Part One: Find directories with accumulated sizes under the size limit.
+  --           Sum up those directories' accumulated sizes (without de-duping the sizes) for the answer.
+  --           Sort the directories by path for sake of display
+  let sizeLimit = 100000
+  let withinLimit = (\(_, size) -> size <= sizeLimit)
+  let indicateWithinLimit = (\entry -> if withinLimit entry then Just " ** SMALL ENOUGH **" else Nothing)
+
+  let (sumWithinLimit, dirsWithinLimit) = accumulateSizes (filter withinLimit dirEntries)
+
+  putStrLn ("~~~~~~~~~~~~~~~~~~ PART ONE ~~~~~~~~~~~~~~~~~~")
+  putStrLn (formatEntries indicateWithinLimit (sortByPath dirEntries))
+  putStrLn ("")
+  putStrLn ("Number of directories within limit:     " ++ show (length dirsWithinLimit))
+  putStrLn ("Total size of directories within limit: " ++ show sumWithinLimit ++ " [PART ONE ANSWER]")
+
+  -- Part Two: Find directories with accumulated sizes above the space needed to free up.
+  --           Then identify the smallest directory that can be deleted to free up enough space.
+  --           To prepare, we first need to calculate how much disk space is used and how
+  --           much space needs to be freed up. We do that by finding the root directory's size.
+  let diskCapacity = 70000000
+  let spaceNeeded  = 30000000
+
+  let totalSizeUsed = maybe 0 snd (find (\(path, size) -> length path == 0) dirEntries)
+  let (freeSpace, spaceToFree) = (diskCapacity - totalSizeUsed, spaceNeeded - freeSpace)
+  let bigEnough = (\(_, size) -> size >= spaceToFree)
+  let indicateBigEnough = (\entry -> if bigEnough entry then Just " ** BIG ENOUGH **" else Nothing)
+
+  let (dirToDelete, spaceFreed) = findSmallest (filter bigEnough dirEntries)
+
+  putStrLn ("")
+  putStrLn ("~~~~~~~~~~~~~~~~~~ PART TWO ~~~~~~~~~~~~~~~~~~")
+  putStrLn (formatEntries indicateBigEnough (sortByPath dirEntries))
+  putStrLn ("")
+  putStrLn ("Total Disk Capacity : " ++ show diskCapacity)
+  putStrLn ("Space Needed        : " ++ show spaceNeeded)
+  putStrLn ("Total Space Used    : " ++ show totalSizeUsed)
+  putStrLn ("Total Space Free    : " ++ show freeSpace)
+  putStrLn ("Space to Free       : " ++ show spaceToFree)
+  putStrLn ("")
+  putStrLn ("Directory to delete : " ++ formatPath dirToDelete)
+  putStrLn ("Space to be Freed   : " ++ show spaceFreed ++ " [PART TWO ANSWER]")
+
+
+findSmallest :: [(path, Int)] -> (path, Int)
+findSmallest = minimumBy (compare `on` snd)
+
+accumulateSizes :: [(path, Int)] -> (Int, [(path, Int)])
+accumulateSizes = mapAccumL (\acc (path, size) -> (acc + size, (path, size))) 0
+
+isDescendent :: ([String], Int) -> Bool
```

Comment:
I believe the correct spelling for "descendent" would be "descend**ant**" ([wiktionary usage notes](https://en.wiktionary.org/wiki/descendant#Usage_notes_2)).